[Enhancement] Configure showing hidden files

benniekiss opened the pull request:
<!--
  Thank you for opening up this pull request! Please read this comment carefully
  and make sure to fill in as much info as possible below as well.

  First, the obligatory checklist. You must make sure to follow this list as
  much as possible to make merging your PR as easy as possible:

  1. I documented all behaviour within the code as far as I could.
  2. This PR targets the develop branch and *not* the master branch.
  3. I have tested my changes extensively and paid extra attention to potential
     cross-platform issues (e./g. Cmd/Ctrl/Super key bindings)
  4. I ran the `yarn lint` and `yarn test` commands successfully
  5. I have added an entry to the CHANGELOG.md
  6. I matched my code-style to the repository as far as possible
  7. I agree that my code will be published under the GNUP GPL v3 license
  8. In case I created new files, I added a header to them (see the existing
     files for examples)
  9. Before opening this PR, I ran `git pull` a last time to prevent any merge
     issues

  If you don't know how to fulfill some of the above points, feel free to open
  your PR nevertheless and mention those points where you are unsure. The more
  you can fulfill, the faster we can merge your PR, otherwise it will just take
  longer.
 -->

## Description
<!-- Below, please describe what the PR does in one or two short sentences. -->
This PR allows a user to configure whether hidden files are shown in the sidebar and file manager.

Closes #6040
Closes #5949

## Changes
<!-- What changes did you make? Please explicitly state any breaking API changes so that nobody is confused why other components suddenly stop working -->
A configuration option was added to the advanced settings pane allowing a user to configure whether hidden files and directories (essentially dot files) are shown in the file manager and sidebar.

Additional extensions configured in the advanced settings pane now show up in the file manager.

Hidden file detection was refactored to be consolidated within the file sorting methods of the `TreeItem`, `FileList`, and `workspace-store`. Now, the FSAL indexing includes all files.

A check was added to the `readDirectoryRecursively` and `readDirectory` FSAL methods to check if the directory is a `.git` dir, in which case parsing is skipped.

`ignoreDir` handling was consolidated, and the FSAL watchdog was updated to use the `ignoreDir` function. The regex to ignore dot files was commented out so that changes to dot files are picked up when a user has them visible.

The `getFilesystemMetadata` function was refactored to be more streamlined and remove references to recursive parsing, which is now no longer handled by this function. The call to `fs.stat` was changed to `fs.lstat` to detect symlinks and prevent crashes when they are in the tree. This led to the target being `stat`'d, but then the subsequent call to `readFile` would fail because it tries to read the link path, not the target path.

## Additional information
<!-- If there is anything else that might be of interest, please provide it here -->

<!-- Please provide any testing system -->
Tested on: <!-- OS including version -->

nathanlesage commented on source/common/util/ignore-dir.ts:
```diff
@@ -16,20 +16,29 @@ import path from 'path'
 // Ignored directory patterns
 import { ignoreDirs } from '../data.json'
```
I believe pulling in those regexps here into this file is a good idea, since they aren't referenced anywhere else. (Also this data file is a remnant from way back, where nothing was consolidated)

nathanlesage commented on source/common/util/file-extention-checks.ts:
```diff
@@ -132,8 +153,7 @@ export function hasDataExt (filePath: string): boolean {
-    hasMSOfficeExt(filePath) || hasDataExt(filePath)
+export function hasAnyRecognizedFileExtension (filePath: string, extensions: string[] = []): boolean {
+  const ext = [ ...ALL_EXT, ...extensions ]
+  return hasExt(filePath, ext)
```
```suggestion
  return hasExt(filePath, [ ...ALL_EXT, ...extensions ])
```

benniekiss commented:
> Okay, so, I think I get this PR. In general, I'd argue we should move from hidden files to dot files, just because it's more accurate and won't confuse Windows users, since there "hiding" files involves setting an attribute.

I had actually originally had this as `dotFiles` and moved to `hiddenFiles` just because there may be other non-dotfile files we would want to hide? But that was just me overthinking the future. I can get all `hiddenFiles` throughout the PR moved back to `dotFiles`.

> Then, it doesn't seem as if this PR actually ignores `.ztr-directory` files, or does it?

No, `.ztr-directory` is not ignored and it is visible when `hiddenFiles` is enabled. Should it be ignored? I can make that happen.

benniekiss commented:
Also, in what layer should the always hidden files be filtered?

Currently, git folders are filtered in the FSAL layer because we dont want to recurse into them, which causes issues when parsing certain "exotic" files within the folder.

However, every other file/folder is filtered at the ui level

danieltomasz commented:
if I can add to the list a quite often folders to exclude can be `.obsidian` and `.quarto` (.git for sure)

benniekiss commented:
Last commit is a WIP to implementing the hard-coded list and filtering at the FSAL level. I believe its 90% there, I just haven't tested extensively.

I renamed the `ignore-dir` file and function to `ignore-path`, etc. because the function is used to check more than just dirs, but rather every file path. I also moved the `isDotFile` function into this file. It felt more at home there, and the previous location in the `file-extension` file was sort of out of place since it was not checking an extension.

Even though we filter at the FSAL level, we still need filtering at the UI level. Since the show dot files config option has two parts -- the file manager and the sidebar -- there may be instances where dotfiles are in the tree but shouldn't be shown in a particular context.

One question I had was how to address watchdog handling. Currently, I re-use the `ignorePath` function in the `chokidar` `ignore` option, and because of that, I believe `ztr-directory` handling is likely broken ATM. What I was wondering -- do we need to have this `ignored` option set, or does the watchdog already ignore the files since we filter them in the FSAL layer?  It is not complicated to have separate watchdog handling, but it means the paths to ignore needs to be separated out. Another question I had was whether we still need to emit events for changes to a `.git` directory -- I couldn't find any clear functionality that depended on these events.

nathanlesage commented:
Everything that is filtered out at the FSAL level also needs to be ignored in the watchdog, especially since changes to git folders would emit a huge amount of events that can have a performance impact on Zettlr because the main loop is busy sending out events into the void.

I just had another question: do we generally need to distinguish between dot files and dot folders?

nathanlesage commented on source/app/service-providers/fsal/index.ts:
```diff
@@ -824,13 +840,20 @@ export default class FSAL extends ProviderContract {
       })
       .map(dirent => {
         const childPath = path.join(directoryPath, dirent.name)
+        if (this.ignorePath(dirent.name)) {
```
This should be a no-op, since you already exclude this in the `filter`, no?

nathanlesage commented on source/common/util/ignore-path.ts:
```diff
@@ -0,0 +1,54 @@
+// A list of path patterns that should be ignored
+// by the FSAL layer and Chokidar watchdog process
+const IGNORE_PATH_RE: RegExp[] = [
+  /(?:^|[\/\\])\.DS_Store$/i, // macOS directory files
```
Can we go back to constructing a list of regexp's like before? This would make adapting the non-changing part of the regexps simpler, and possibly make the list of file types we ignore easier to ignore, too?

nathanlesage commented on source/common/util/ignore-path.ts:
```diff
@@ -0,0 +1,54 @@
+ * @return  {boolean}            Whether the file is a dot file or folder
+ */
+export function isDotFile (filePath: string): boolean {
+  return /(?:^|[\/\\])\./.test(filePath)
```
Wouldn't it be safer to use `path.basename().startsWith('.')`?

benniekiss commented on source/app/service-providers/fsal/index.ts:
```diff
@@ -124,7 +124,7 @@ export default class FSAL extends ProviderContract {
 
     this._config.on('update', (which: string) => {
-      if (which === 'openPaths') {
+      if (which === 'openPaths' || which === 'files.dotFiles.showInFilemanager' || which === 'files.dotFiles.showInSidebar') {
```
No, this is just a trigger to re-index the files, no? Since we exclude dotfiles at the FSAL level, when we enable/disable the option to show them, we need to reindex to capture them.

benniekiss commented on source/win-main/file-manager/FileList.vue:
```diff
@@ -155,41 +164,51 @@ const getDirectoryContents = computed<RecycleScrollerData[]>(() => {
-        }
-      } else if (hasDataExt(sortedDescendants[i].path)) {
-        if (!files.dataFiles.showInFilemanager) {
+    if (sortedDescendants[i].type === 'directory') {
```
I moved this to a filter/map, which is a little more streamlined

benniekiss commented:
> Given that Zettlr may need to react to those changes, we should leave it. But anything that Zettlr doesn't need should be ignored. If we have to extend the list of exceptions in the future so be it!

So I did some testing to see what was emitted for git dirs, and I don't think anything relies on those types of event, and I don't think the code actually picks them up anyways. In a separate terminal, I tried doing some git commands on a folder opened as a workspace in Zettlr, and no events were emitted, doing things like `touch test.md; git add test.md; git commit -m 'test'`, etc. did not emit the `.git` watchdog events. I think it should be safe to remove this special handling, as I don't think its needed, and I'm not sure its functional.

Regarding `.ztr-directory`, though, that does have events appropriately picked up, and I think it is used by the Zettlr for monitoring, so that should stay in.

benniekiss commented:
Last commit builds the regex programmatically, which I think addresses all your points. I believe this is good to go on my end

nathanlesage commented on source/common/util/ignore-path.ts:
```diff
@@ -0,0 +1,89 @@
+
+// Matches either the end of the line
+// or an internal segment of the path.
+const re_suffix = '(?:$|[\\\/\\\\])'
```
```suggestion
const suffixRE = '(?:$|[\\\/\\\\])'
```

nathanlesage commented on source/app/service-providers/fsal/fsal-directory.ts:
```diff
@@ -128,34 +136,22 @@ export async function parse (currentPath: string): Promise<DirDescriptor> {
+  try {
+    dir.isGitRepository = (await fs.lstat(path.join(dir.path, '.git'))).isDirectory()
+  } catch (err: any) {}
+
```
Can't we do a simple `isDir`-check? That would avoid this try/catch contraption.

nathanlesage commented on source/app/service-providers/fsal/fsal-watchdog.ts:
```diff
@@ -61,10 +43,10 @@ export default class FSALWatchdog {
       persistent: true,
       ignoreInitial: true, // Do not track the initial watch as changes
-      followSymlinks: true, // Follow symlinks
+      followSymlinks: false, // Zettlr does not support symlinks
```
This is a beautiful diff to read.

nathanlesage commented:
Thanks a bunch! Looks good now – merging
